drivers/gpio/gpio_pca_series.c

```diff
@@ -0,0 +1,2169 @@
+	struct gpio_callback gpio_cb; /** gpio_int ISR callback */
+	sys_slist_t callbacks;        /** port pin callbacks list */
+	struct k_work int_work;       /** worker that fire callbacks */
+#endif
+};
+
+/**
+ * gpio_pca_reg_access_api
+ * {
+ */
+
+static inline uint8_t gpio_pca_series_reg_get_addr(const struct device *dev,
+						   enum gpio_pca_series_reg_type reg_type)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+
+	return cfg->part_cfg->regs[reg_type];
+}
+
+/**
+ * @brief get read size for register
+ *
+ * @param dev
+ * @param reg_type
+ * @return uint32_t size in bytes
+ *                  0 if fail
+ */
+static inline uint32_t gpio_pca_series_reg_get_size(const struct device *dev,
+							enum gpio_pca_series_reg_type reg_type)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+
+	switch (reg_type) {
+	case PCA_REG_TYPE_1B_INPUT_PORT:
+	case PCA_REG_TYPE_1B_OUTPUT_PORT:
+/*  case PCA_REG_TYPE_1B_POLARITY_INVERSION: */
+	case PCA_REG_TYPE_1B_CONFIGURATION:
+	case PCA_REG_TYPE_1B_INPUT_LATCH:
+	case PCA_REG_TYPE_1B_PULL_ENABLE:
+	case PCA_REG_TYPE_1B_PULL_SELECT:
+	case PCA_REG_TYPE_1B_INPUT_STATUS:
+	case PCA_REG_TYPE_1B_OUTPUT_CONFIG:
+#ifdef CONFIG_GPIO_PCA_SERIES_INTERRUPT
+	case PCA_REG_TYPE_1B_INTERRUPT_MASK:
+	case PCA_REG_TYPE_1B_INTERRUPT_STATUS:
+	case PCA_REG_TYPE_1B_INTERRUPT_CLEAR:
+#endif /* CONFIG_GPIO_PCA_SERIES_INTERRUPT */
+		return cfg->part_cfg->port_no;
+	case PCA_REG_TYPE_2B_OUTPUT_DRIVE_STRENGTH:
+#ifdef CONFIG_GPIO_PCA_SERIES_INTERRUPT
+	case PCA_REG_TYPE_2B_INTERRUPT_EDGE:
+#endif /* CONFIG_GPIO_PCA_SERIES_INTERRUPT */
+		return (cfg->part_cfg->port_no * 2U);
+	default:
+		return 0; /** should never happen */
+	}
+}
+
+#ifdef CONFIG_GPIO_PCA_SERIES_CACHE_ALL
+/** forward declarations */
+static inline int gpio_pca_series_reg_update_cache(struct device *dev,
+						   enum gpio_pca_series_reg_type reg_type,
+						   uint8_t *buf);
+#endif /* CONFIG_GPIO_PCA_SERIES_CACHE_ALL */
+
+/**
+ * @brief read register with i2c interface.
+ *
+ * @note if CONFIG_GPIO_PCA_SERIES_CACHE_ALL is enabled, this will
+ *          also update reg cache.
+ *
+ * @param dev
+ * @param reg_type
+ * @param buf pointer to read data
+ * @return int 0        if success
+ *             -EFAULT  if register is not supported
+ *             -EIO     if i2c failure
+ */
+static inline int gpio_pca_series_reg_read(const struct device *dev,
+					   enum gpio_pca_series_reg_type reg_type, uint8_t *buf)
+{
+	const struct gpio_pca_series_config *cfg = dev->config;
+	int ret;
+	uint32_t size = gpio_pca_series_reg_get_size(dev, reg_type);
+	uint8_t addr = gpio_pca_series_reg_get_addr(dev, reg_type);
+
+	LOG_DBG("reg read type %d addr 0x%x len %d", reg_type, addr, size);
+
+	if (addr == PCA_REG_INVALID) {
+		LOG_ERR("trying to read unsupported reg, reg type %d", reg_type);
+		return -EFAULT;
+	}
+
+	ret = i2c_burst_read_dt(&cfg->i2c, addr, buf, size);
+	if (ret) {
+		LOG_ERR("i2c read error [%d]", ret);
+		return ret;
+	}
```

Comment:
Switched back to burst_read api **temporarily**, because using `i2c_transfer` directly is causing i2c transaction to hang. Not sure why, will debug later.